eradiate/scenes/biosphere/_discrete.py

```diff
@@ -947,18 +1066,161 @@ def bsdfs(self, ctx=None):
+            dictionary generation.
+
+        Returns → dict:
+            Return a dictionary suitable for merge with a :class:`.KernelDict`
+            containing all the BSDFs attached to the shapes in the leaf cloud.
+        """
+
+        bsdfs_dict = self.leaf_cloud.bsdfs(ctx=ctx)
+
+        bsdfs_dict[f"bsdf_{self.id}"] = {
+            "type": "diffuse",
+            "reflectance": self.trunk_reflectance.kernel_dict(ctx=ctx)["spectrum"],
+        }
+
+        return bsdfs_dict
+
+    def shapes(self, ctx=None):
+        """
+        Return shape plugin specifications.
+
+        Parameter ``ctx`` (:class:`.KernelDictContext` or None):
+            A context data structure containing parameters relevant for kernel
+            dictionary generation.
+
+        Returns → dict:
+            A dictionary suitable for merge with a
+            :class:`~eradiate.scenes.core.KernelDict` containing all the shapes
+            in the canopy.
+        """
+        from mitsuba.core import ScalarTransform4f
+
+        kernel_length = uck.get("length")
+
+        kernel_height = self.trunk_height.m_as(kernel_length)
+        kernel_radius = self.trunk_radius.m_as(kernel_length)
+
+        leaf_cloud = self.leaf_cloud.translated([0., 0., kernel_height]*kernel_length)
+
+        if ctx.ref:
+            bsdf = {"type": "ref", "id": f"bsdf_{self.id}"}
         else:
-            return {**self.bsdfs(ctx=ctx), **self.shapes(ctx=ctx)}
+            bsdf = self.bsdfs(ctx=ctx)[f"bsdf_{self.id}"]
+
+        shapes_dict = leaf_cloud.shapes(ctx=ctx)
+
+        shapes_dict[f"trunk_cyl_{self.id}"] = {
+            "type": "cylinder",
+            "bsdf": bsdf,
+            "radius": kernel_radius,
+            "p0": [0, 0, -0.1],
+            "p1": [0, 0, kernel_height],
+        }
+
+        shapes_dict[f"trunk_cap_{self.id}"] = {
+            "type": "disk",
+            "bsdf": bsdf,
+            "to_world": ScalarTransform4f.scale(kernel_radius)
+            * ScalarTransform4f.translate(((0, 0, kernel_height / 2.0))),
+        }
+
+        return shapes_dict
+
+    @staticmethod
+    def convert(value):
+        """
+        Object converter method.
+
+        If ``value`` is a dictionary, this method uses :meth:`from_dict` to
+        create a :class:`.CanopyElement`.
+
+        Otherwise, it returns ``value``.
+        """
+        if isinstance(value, dict):
+            return AbstractTree.from_dict(value)
+
+        return value
+
+    @classmethod
+    def from_dict(cls, d):
+        """
+        Construct from a dictionary.
+
+
+        Parameter ``d`` (dict):
+            Dictionary containing parameters passed to the selected constructor.
+            Unit fields are pre-processed with :func:`pinttr.interpret_units`.
+        """
+
+        # Interpret unit fields if any
+        d_copy = pinttr.interpret_units(d, ureg=ureg)
+
+        leaf_cloud_dict = d_copy.pop("leaf_cloud")
+
+        leaf_cloud = LeafCloud.convert(leaf_cloud_dict)
```

Comment:
Did you have the opportunity to check this?

Reply:
Your approach would lead to a name collision in the `AbstractTree` constructor. On one hand I pass the instantiated leaf cloud from Line 1233 as the `leaf_cloud` argument but the `*d_copy` would add another `**kwarg` with the name `leaf_cloud` :thinking: